### DISCUSSION TITLE
1주차 사전 과제, 김우민

### DISCUSSION THREAD

**Discussion by gleaming9** [MrMirror21/2024-KHU-CSE-TrackStudy-Web#19] [General]
### 1. 명명 규칙(Naming Convention)이라는 것이 있습니다. 어떤 규칙들이 있는지, 각 규칙들은 어떻게 코드를 작성하는지, 자신이 그동안 사용해왔던 명명 규칙은 무엇인지 적어주세요.

- 카멜 케이스 (Camel Case)
Lower camel case: 첫 번째 단어를 제외한 각 단어의 첫 글자를 대문자로 시작한다. 예: firstName, speedOfLight
Upper camel case (Pascal Case): 모든 단어의 첫 글자를 대문자로 시작한다. 예: ProductInfo, CarEngine

- 스네이크 케이스 (Snake Case)
모든 단어를 소문자로 작성하고, 단어 사이를 언더스코어(_)로 연결한다. 예: record_status, last_update_date

- 키밥 케이스 (Kebab Case)
모든 단어를 소문자로 작성하고, 단어 사이를 하이픈(-)으로 연결한다. 주로 HTML과 CSS에서 자주 사용한다. 예: menu-item, background-color

- 헝가리안 케이스 (Hungarian Notation)
변수나 함수의 이름 앞에 타입을 미리 지정하는 접두어를 붙여 타입을 명시한다. 예: strName, bIsActive

- 대문자 스네이크 케이스
모든 글자를 대문자로 쓰고, 단어 사이는 언더스코어로 연결한다. 주로 상수를 명명할 때 사용한다. 예: MAX_HEIGHT, DEFAULT_COLOR

- 개인적 경험
대부분 스네이크 케이스를 사용해서 변수명을 정했던 것 같다. 클래스의 경우에는 카멜케이스를 활용했다.

### 2. 함수가 순수하다는 것은 무엇을 의미할까요?
 특정한 조건을 만족하는 함수의 성질을 말한다.
- 참조 투명성
동일한 입력에 대해서 항상 동일한 출력을 반환한다. 순수 함수는 함수의 인자에 의해서만 결과가 결정된다. 외부 상태나 전역변수를 활용하지 않는다.
- No Side effects
외부 상태를 변경하지 않는다.  함수의 실행이 외부 상태나 함수 외부의 다른 변수, 데이터 구조, I/O 등에 영향을 미치지 않는다. 함수 내부에서만 작업이 수행된다.

### 3. 개발 시에 순수 함수(pure function)로만 모든 코드를 작성하는 것에 대한 생각을 적어주세요.
순수 함수만으로 모든 코드를 작성하는 것은 이상적이지만 현실적으로 모든 프로그램이 순수 함수만으로 구성될 수는 없다고 생각한다. 대신 가능한 많은 부분을 순수 함수로 작성하고, 불가피하게 순수 함수를 사용할 수 없는 부분은 명확하게 분리하여 관리하면 좋을 것 같다.

### 4. 자바스크립트에서 함수를 작성하는 문법은 다음과 같습니다. 이 외의 함수를 작성하는 문법 두 가지가 무엇이며, 어떤 때 사용하나요?
```
function myFunc(param){
    statements
}
```
- 함수 표현식
```
const myFunc = function(param){
    //statements
}
```
함수 표현식은 함수를 변수에 할당하는 방식으로 정의한다. 익명 함수로 정의가 가능하다. 함수 선언과 달리 코드 블록(ex: if문, for문)에서도 정의될 수 있다. 함수 선언과 달리, 함수 표현식은 호이스팅되지 않는다. 즉, 함수가 정의되기 전에 호출할 수 없다.
익명 함수 형태로 주로 콜백 함수로 사용한다.

- 화살표 함수
화살표 함수는 ES6에서 도입된 문법으로, 보다 간결한 함수 표현식을 제공한다.
```
const myFunc = (param) => {
    //statements
}
// 매개변수가 하나일 때 괄호 생략 가능
const myFunc = param =>{ 
    //statements
}
//한 줄로 반환할 때 중괄호와 return 생략 가능
const myFunc = param => param*2;
```
화살표 함수는 자신만의 'this'바인딩을 가지지 않는다. 즉, 화살표 함수 내부의 this는 상위 스코프의 this를 가리킨다. 예를 들어, 객체 메소드 내에서 내부 함수가 this로 상위 객체를 참조해야 할 때 유용하다. 콜백 함수로 자주 사용된다.

### 5. 파이썬에서는 문자열을 포맷팅하기 위해 다음과 같은 f-string을 사용합니다. 자바스크립트에서 f-string과 같이 문자열을 다양한 자료형과 변수들을 혼합하여 사용하려면 어떤 방법을 사용하나요?
```
>>> x = 1
>>> y = 2
>>> f"{x} + {y}는 {x + y}입니다."
'1 + 2는 3입니다.'
```
- Temlate literals, backtik
문자열을 편리하게 작성하고 다룰 수 있도록 고안된 문법이다. 기존의 문자열 표기방식인 ', " 대신 backtik `을 사용하여 문자열을 감싸는 것이 주요 특징이다. 일반 문자열과 달리 여러 줄에 걸쳐 작성할 수 있으며, 변수와 표현식을 삽입할 수 있어서 문자열 조작을 더욱 편리하게 할 수 있다.
```
var x = 1;
var y = 2;
var message = `${x} + ${y}는 {x + y}입니다.`;
console.log(message);
```

### 6. 다음과 같은 객체(object)에서 p 값을 출력하려면 o.p 와 같은 방식으로 출력해야 하는데, p 로 출력할 수 있는 방법은 없을까요?
```
var o = { p: 42, q: true };

console.log(o.p); // 42
console.log(p); // p is undefined
```
- 객체의 속성을 전역 변수로 만들기
```
var o = { p: 42, q: true };
var p = o.p;

console.log(p); // 42
```
- Destructuring Assignment
ES6에서 도입된 디스트럭처링을 사용하면 객체의 속성을 개별 변수로 쉽게 할당이 가능하다.
```
var o = { p: 42, q: true };

// 디스트럭처링을 사용하여 객체 속성을 개별 변수로 할당
var { p, q } = o;

console.log(p); // 42
console.log(q); // true
```
Desturcturing assignment를 사용하면 객체의 특정 속성을 간단하게 변수로 추출할 수 있어 가독성과 유지보수성이 높아진다는 장점을 가지고 있다.

**Comment by MrMirror21** [comment]
답변해주신 사전 과제 잘 봤습니다.
전체적으로 퀄리티 높게 아주 잘 답변해주셨네요 👍
앞으로도 수업 내용 이해에 도움이 되는 사전 과제 최대한 잘 준비해올테니 열심히 참여해주시기 바랍니다.
